File: internal/analyzer/detectors/suspicious_protocol.py
from typing import Dict, List, Set, Tuple

import pandas as pd

from internal.analyzer.base import Alert, BaseDetector, Level


class SuspiciousProtocolDetector(BaseDetector):
    def __init__(self):
        super().__init__()
        self.common_ports = {
            80: ["TCP", "HTTP"],
            443: ["TCP", "HTTPS"],
            53: ["UDP", "TCP", "DNS"],
            22: ["TCP", "SSH"],
            25: ["TCP", "SMTP"],
            5353: ["UDP", "mDNS"],
        }
        self.suspicious_mappings = {
            80: ["SSH"],
            22: ["UDP"],
        }

    def analyze(self, df: pd.DataFrame) -> List[Alert]:
        alerts = []
        if (
            df.empty
            or "src_ip" not in df.columns
            or "dst_port" not in df.columns
            or "protocol" not in df.columns
        ):
            return alerts

        clean_df = df.dropna(subset=["src_ip", "dst_port", "protocol"])
        clean_df = clean_df[clean_df["src_ip"].str.strip() != ""]

        if clean_df.empty:
            return alerts

        for _, row in clean_df.iterrows():
            ip = row["src_ip"]
            port = int(row["dst_port"])
            proto = str(row["protocol"]).upper()

            is_suspicious = False
            reason = ""

            if (
                port in self.suspicious_mappings
                and proto in self.suspicious_mappings[port]
            ):
                is_suspicious = True
                reason = f"Unusual protocol {proto} on port {port}"

            elif port in self.common_ports and proto not in self.common_ports[port]:
                pass

            if is_suspicious:
                if self._should_alert("SUSPICIOUS_PROTOCOL", ip):
                    alerts.append(
                        Alert(
                            timestamp=str(row["timestamp"]),
                            level=Level.WARNING,
                            type="SUSPICIOUS_PROTOCOL",
                            message=f"[{ip}] {reason}",
                            src_ip=ip,
                        )
                    )

        return alerts

File: internal/analyzer/detectors/suspicious_protocol.go
package detectors

import (
	"fmt"
	"hattic/internal/analyzer"
	"strings"
	"time"
)

type SuspiciousProtocolDetector struct {
	*analyzer.BaseDetector
	commonPorts        map[uint16]map[string]bool
	suspiciousMappings map[uint16]map[string]bool
}

func NewSuspiciousProtocolDetector() *SuspiciousProtocolDetector {
	d := &SuspiciousProtocolDetector{
		BaseDetector:       analyzer.NewBaseDetector(5 * time.Minute),
		commonPorts:        make(map[uint16]map[string]bool),
		suspiciousMappings: make(map[uint16]map[string]bool),
	}

	d.commonPorts[80] = map[string]bool{"TCP": true, "HTTP": true}
	d.commonPorts[443] = map[string]bool{"TCP": true, "HTTPS": true}
	d.commonPorts[53] = map[string]bool{"UDP": true, "TCP": true, "DNS": true}
	d.commonPorts[22] = map[string]bool{"TCP": true, "SSH": true}
	d.commonPorts[25] = map[string]bool{"TCP": true, "SMTP": true}
	d.commonPorts[5353] = map[string]bool{"UDP": true, "MDNS": true}

	d.suspiciousMappings[80] = map[string]bool{"SSH": true}
	d.suspiciousMappings[22] = map[string]bool{"UDP": true}

	return d
}

func (d *SuspiciousProtocolDetector) Analyze(packet *analyzer.PacketInfo) []analyzer.Alert {
	var alerts []analyzer.Alert

	if packet.SrcIP == "" || packet.DstPort == 0 || packet.Protocol == "" {
		return alerts
	}

	proto := strings.ToUpper(packet.Protocol)
	isSuspicious := false
	reason := ""

	if protos, ok := d.suspiciousMappings[packet.DstPort]; ok && protos[proto] {
		isSuspicious = true
		reason = fmt.Sprintf("Unusual protocol %s on port %d", proto, packet.DstPort)
	}

	if isSuspicious {
		if d.ShouldAlert("SUSPICIOUS_PROTOCOL", packet.SrcIP) {
			alerts = append(alerts, analyzer.Alert{
				Timestamp: packet.Timestamp.Format(time.RFC3339),
				Level:     analyzer.LevelWarning,
				Type:      "SUSPICIOUS_PROTOCOL",
				Message:   fmt.Sprintf("[%s] %s", packet.SrcIP, reason),
				SrcIP:     packet.SrcIP,
			})
		}
	}

	return alerts
}
